daml-lf/spec/contract-id.rst

```diff
@@ -0,0 +1,245 @@
+.. Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+.. SPDX-License-Identifier: Apache-2.0
+
+V1 Contract ID allocation scheme
+================================
+
+Goals
+^^^^^
+
+* Allows ordering contract IDs and make DAML semantics depend on this
+  order, e.g., for comparison builtin and maps using IDs as keys.
+* Eliminate all contract ID translations for central commiter ledger
+
+Requirements
+^^^^^^^^^^^^
+
+**Executability**: DAML engine can determine ordering of the contract
+IDs before it hands out the transaction to the write service.
+
+**Validation**: The allocation scheme commutes with transaction
+projection. That is, the contract IDs for contracts created in the
+projection of a transaction to a set of parties can be computed
+solely from the projection and input seeds.
+
+**Unlinkability**: It is computationally infeasible to link the contract
+contents to the contract ID unless the create node is witnessed. The
```

Comment:
As is this is too strong. Given that transactions are small in size, an adversary that knows just the input seed used for the transaction that creates a contract could derive seeds for all potential nodes in the tree and then validate its guess of the signatories by recomputing the discriminators.
```suggestion
contents to the contract ID unless the create node is witnessed or the input seeds are known. The
```